Cache controller

peyuaa opened the pull request:
A controller that, when the cache is full, iterates over all cache items and removes all expired

peyuaa committed:
moved file paths to constants

peyuaa committed:
started work on cache controller that will used to delete expired cache

peyuaa committed:
redundant break statement

peyuaa committed:
added error check

peyuaa committed:
moved error check closer to function

peyuaa committed:
reorganized imports

peyuaa committed:
mvp cacheController

peyuaa committed:
run cache controller in separate goroutine

peyuaa committed:
Merge branch 'master' into cache-clearing

peyuaa committed:
deleted unused errors import

peyuaa committed:
added log message when cache controller has been started

peyuaa committed:
added anonimous functions in deleteExpiredCache to made it mode readable

pelageech commented on cache_control/cacheController.go:
```diff
@@ -0,0 +1,82 @@
+	frequency   *time.Ticker
+}
+
+func New(db *bolt.DB, dbFile *os.File, maxFileSize int64, frequency *time.Ticker) *cacheController {
```
Предупреждение о том, что возвращется неэкспортируемый тип в экспортируемой функции. Это норм?

peyuaa replied:
Я поменял на экспортируемый, потому что экспортируемые методы неэкспортируемой структуры не показываются в godoc. Надо написать комментарии к функции observe

pelageech commented on cache_control/cacheController.go:
```diff
@@ -0,0 +1,82 @@
+	}
+}
+
+func (c *cacheController) isExpired(info cache.Info) bool {
```
Очень хорошо бы в read такую функцию заюзать.

peyuaa replied:
I need more details....

В cache/read.go?

pelageech replied:
yes:)

Commit:
DeleteRecord was corrected

Commit:
Added a constant 'info' that is convenient to work with cache. For each bucket there will be info key

Commit:
remove last "/" in root.
all the functions working with disk are corrected because of the reason above.
Added RemovePageFromDisk function

Commit:
put removing from disk to cacheController

Commit:
refactor: info renamed to pageInfo

pelageech commented on cache_control/cacheController.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+func (c *cacheController) isSizeExceeded() bool {
+	fileInfo, err := c.dbFile.Stat()
```
Я правильно понял, что здесь размер именно базы данных берётся?
Я на самом деле хотел, чтобы по размеру физических файлов на диске сравнивалось.

peyuaa replied:
А почему у нас вообще файлы хранятся отдельно на диске, а не в той же базе данных?

peyuaa replied:
Да, здесь размер именно базы данных

pelageech replied:
Потому что болт хранится в памяти, а Иртегов сказал, что это дикость хранить страницы в памяти. В памяти хранятся только метаданные

pelageech commented on cache_control/cacheController.go:
```diff
@@ -0,0 +1,88 @@
+		log.Printf("Error getting file info in cacheController: %v", err) //todo: how to handle this error properly?
+	}
+
+	return fileInfo.Size() > c.maxFileSize
```
Надо на какую-нибудь константу умножить c.maxFileSize, чтобы он, например, не пытался больше 90% заполниться

peyuaa replied:
Можно. Ну или можно сразу в c.maxFileSize писать необходимую загруженность в 90%, а не 100%.

Если честно неочевидно, что лучше, энивей не сегодня буду все это смотреть скорее всего

pelageech replied:
Думаю, две константы норм будет)
Одна контролирует размер диска, другая критический занимаемый размер

Commit:
Added size into cache.Info

Commit:
refactor

peyuaa committed:
made struct exported so that exported methods of that struct will be shown in godoc

peyuaa committed:
Merge branch 'cache-clearing' of github.com:pelageech/BDUTS into cache-clearing

peyuaa committed:
changed dbFiles root and now check size of the db files, written to disk, not a bolt files with metadata

peyuaa committed:
added fillFactor of the database. After database files sizes exceeds fillFactor*maxDBSize we're starting to delete expired cache

peyuaa committed:
fixed 404 error

peyuaa committed:
create database files directory if not exist
